scikeras/utils/transformers.py

```diff
@@ -0,0 +1,189 @@
+                " label type."
+            )
+        self.final_encoder_ = encoders[target_type].fit(y)
+
+        if (
+            target_type == "multilabel-indicator"
+            and y.min() == 0
+            and (y.sum(axis=1) == 1).all()
+        ):
+            target_type = "multiclass-onehot"
+
+        self.n_outputs_ = 1
+        self.model_n_outputs_ = 1
+        self.y_dtype_ = y.dtype
+        self._target_type = target_type
+
+        if target_type in ("binary", "multiclass"):
+            self.classes_ = self.final_encoder_[1].categories_[0]
+            self.n_classes_ = self.classes_.size
+        elif target_type in ("multiclass-onehot", "multilabel-indicator"):
+            self.classes_ = np.arange(0, y.shape[1])
+            self.n_classes_ = y.shape[1]
+        elif target_type == "multiclass-multioutput":
+            self.classes_ = [np.unique(y[:, j]) for j in range(y.shape[1])]
+            self.n_classes_ = len(self.n_classes_)
+
+        return self
+
+    def transform(self, X: np.ndarray) -> np.ndarray:
+        y = X  # rename for clarity, the input is always expected to be a target `y`
+        # no need to validate n_outputs_ or model_n_outputs_, those are hardcoded
+        # self.classes_ and self.n_classes_ are validated by the transformers themselves
+        return self.final_encoder_.transform(y)
+
+    def inverse_transform(
+        self, X: np.ndarray, return_proba: bool = False
+    ) -> np.ndarray:
+        y = X  # rename for clarity, the input is always expected to be a target `y`
+        if self._target_type == "binary":
+            # array([0.9, 0.1], [.2, .8]) -> array(['yes', 'no'])
+            if y.ndim == 1 or (y.shape[1] == 1 and self.n_classes_ == 2):
+                # result from a single sigmoid output
+                # reformat so that we have 2 columns
+                y = np.column_stack([1 - y, y])
+            y_ = np.argmax(y, axis=1).reshape(-1, 1)
+            class_predictions = self.final_encoder_.inverse_transform(y_)
+        elif self._target_type == "multiclass":
+            # array([0.8, 0.1, 0.1], [.1, .8, .1]) ->
+            # array(['apple', 'orange'])
+            idx = np.argmax(y, axis=-1)
+            if not is_categorical_crossentropy(self.loss):
+                y_ = idx.reshape(-1, 1)
+            else:
+                y_ = np.zeros(y.shape, dtype=int)
+                y_[np.arange(y.shape[0]), idx] = 1
+            class_predictions = self.final_encoder_.inverse_transform(y_)
+        elif self._target_type == "multiclass-onehot":
+            # array([.8, .1, .1], [.1, .8, .1]) ->
+            # array([[1, 0, 0], [0, 1, 0]])
+            idx = np.argmax(y, axis=-1)
+            y_ = np.zeros(y.shape, dtype=int)
+            y_[np.arange(y.shape[0]), idx] = 1
+            class_predictions = y_
+        elif self._target_type in ("multilabel-indicator", "multiclass-multioutput"):
+            class_predictions = None
```

Comment:
It's a bit unclear to me what to do here. For `"multilabel-indicator"` I think we can do `class_predictions = np.around(y)`, but I need to convince myself of it. For `multiclass-multiouput` I don't think there's anything universal.

Reply:
Should `inverse_transform(transform(y)) == y`?

Reply:
Generally I think so, but I realize that that's not the case for the KerasClassifier transformers. This is so that `predict` can call `inverse_transform` with no arguments and get back raw predictions for regressors and classes for classifiers. Otherwise we would have to override the entire `KerasClassifier.predict` method to change a single argument. But thinking about it that does sound cleaner from an interface perspective, even if it results in added complexity within `wrappers.py`.

Reply:
I think if the user passes in data of a particular flavor, they should get the same flavor in return. That means I think an input of onehot encoded data should result in an output of onehot encoded data. Likewise, I think an input of a 1D array of ints should result in a 1D array of ints as output.

Reply:
Yeah I agree. I think maybe I had it the wrong way round in https://github.com/adriangb/scikeras/pull/88#discussion_r502996245. By having `predict_proba=False` as the default (as we currently have), `inverse_transform(transform(y)) == y` is valid, right? Which looping back to the first comment in this thread means we _should_ implement `np.around(y)` for `multilabel-indicator`.

Unless I'm missing the relation of https://github.com/adriangb/scikeras/pull/88#discussion_r502992500 to https://github.com/adriangb/scikeras/pull/88#discussion_r502987550?

Reply:
I thought about this a bit more. I think everything I said above is wrong.

`fit` takes labels but `predict` returns probability. Since we feed `transform -> fit` and `predict -> inverse_transform`, `inverse_transform(transform(y)) != y`. I don't see any way to change this that doesn't involve a lot of complexity.

Reply:
How does Scikit-learn's [MLPClassifier][1] behave when predicting on one-hot data?

> predict returns probability

Does it? There are two prediction functions for MLPClassifier; `predict` and `predict_proba`.

[1]:https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html

Reply:
> predict returns probability

I was specifically referring to our `BaseWrapper.predict`. For Classifiers, `predict` should return class predictions, which in our case is achieved by having the default value for `return_proba = False`, allowing us to re-use the implementation in `BaseWrapper.predict`. Of course we can flip it (make `return_proba = True` the default) and have `def predict_proba(...): super().predict(...)` and also override `predict` to add `return_proba=False`.

Reply:
> How does Scikit-learn's [MLPClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html) behave when predicting on one-hot data?

```python3
from sklearn.datasets import make_classification
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import OneHotEncoder


X, y = make_classification()
tf = OneHotEncoder(sparse=False)
y_tf = tf.fit_transform(y.reshape(-1, 1))
clf = MLPClassifier()
clf.fit(X, y_tf)
print(clf.predict(X)[:5])
```
```
[[1 0]
 [0 1]
 [0 1]
 [0 1]
 [0 1]]
```

Reply:
Do we want to do something here @stsievert ?